presto-main/src/test/java/com/facebook/presto/metadata/TestSignatureBinder.java

```diff
@@ -986,11 +988,82 @@ public void testUnknownCoercion()
+                .returnType(parseTypeSignature("V"))
+                .argumentTypes(parseTypeSignature("T"), parseTypeSignature("function(T,U)"), parseTypeSignature("function(U,V)"))
+                .typeVariableConstraints(typeVariable("T"), typeVariable("U"), typeVariable("V"))
+                .build();
+        assertThat(applyTwice)
+                .boundTo("integer", "function(integer,varchar)", "function(varchar,double)")
+                .produces(BoundVariables.builder()
+                        .setTypeVariable("T", INTEGER)
+                        .setTypeVariable("U", VARCHAR)
+                        .setTypeVariable("V", DOUBLE)
+                        .build());
+        assertThat(applyTwice)
+                .boundTo(
+                        "integer",
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(integer,varchar)")),
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(varchar,double)")))
+                .produces(BoundVariables.builder()
+                        .setTypeVariable("T", INTEGER)
+                        .setTypeVariable("U", VARCHAR)
+                        .setTypeVariable("V", DOUBLE)
+                        .build());
+        assertThat(applyTwice)
+                .boundTo(
+                        // pass function argument to non-function position of a function
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(integer,varchar)")),
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(integer,varchar)")),
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(varchar,double)")))
+                .fails();
+        assertThat(applyTwice)
+                .boundTo(
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(integer,varchar)")),
+                        // pass non-function argument to function position of a function
+                        "integer",
+                        new TypeSignatureProvider(functionArgumentTypes -> TypeSignature.parseTypeSignature("function(varchar,double)")))
+                .fails();
+
+        Signature flatMap = functionSignature()
+                .returnType(parseTypeSignature("array(T)"))
+                .argumentTypes(parseTypeSignature("array(T)"), parseTypeSignature("function(T, array(T))"))
+                .typeVariableConstraints(typeVariable("T"))
+                .build();
+        assertThat(flatMap)
+                .boundTo("array(integer)", "function(integer, array(integer))")
+                .produces(BoundVariables.builder()
+                        .setTypeVariable("T", INTEGER)
+                        .build());
+    }
```

Comment:
How about varargs? I'd test a function `apply_all(T, function<T, T>...) -> T` and make sure it can bind to `apply_all(1, x -> x + 1)` and `apply_all(1, x -> x + 1, y -> y + 1)`